File: 654.最大二叉树.cs
/*
 * @lc app=leetcode.cn id=654 lang=csharp
 *
 * [654] 最大二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    public TreeNode ConstructMaximumBinaryTree(int[] nums)
    {
        if (nums.Length == 0) return null;

        return MaximumBinaryTree(nums, 0, nums.Length - 1);
    }

    public TreeNode MaximumBinaryTree(int[] nums, int left, int right)
    {
        if (left > right) return null;

        int maxValue = Int32.MinValue;
        int maxIndex = -1;
        for (int i = left; i < right + 1; ++i)
        {
            if (nums[i] > maxValue)
            {
                maxValue = nums[i];
                maxIndex = i;
            }
        }

        TreeNode root = new TreeNode(maxValue);
        root.left = MaximumBinaryTree(nums, left, maxIndex - 1);
        root.right = MaximumBinaryTree(nums, maxIndex + 1, right);

        return root;
    }
}
// @lc code=end

File: 654.最大二叉树.cpp
/*
 * @lc app=leetcode.cn id=654 lang=cpp
 *
 * [654] 最大二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode *constructMaximumBinaryTree(vector<int> &nums) {
        int n = nums.size();
        stack<int> stk;
        vector<int> l(n, -1), r(n, -1);
        vector<TreeNode *> trees(n);
        for (int i = 0; i < nums.size(); ++i) {
            trees[i] = new TreeNode(nums[i]);
            int num = nums[i];
            while (!stk.empty() && num > nums[stk.top()]) {
                r[stk.top()] = i;
                stk.pop();
            }
            l[i] = stk.empty() ? -1 : stk.top();
            stk.emplace(i);
        }

        TreeNode *root = nullptr;
        for (int i = 0; i < n; ++i) {
            if (l[i] == -1 && r[i] == -1) {
                root = trees[i];
            } else if (l[i] == -1 || (r[i] != -1 && nums[r[i]] < nums[l[i]])) {
                trees[r[i]]->left = trees[i];
            } else {
                trees[l[i]]->right = trees[i];
            }
        }
        return root;
    }
};
// @lc code=end
